ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,255 @@
+    #else
+        if (type != ULISTFMT_TYPE_AND) {
+            zend_argument_value_error(2, "ICU 66 and below only support IntlListFormatter::TYPE_AND");
+            RETURN_THROWS();
+        }
+
+        if (width != ULISTFMT_WIDTH_WIDE) {
+            zend_argument_value_error(3, "ICU 66 and below only support IntlListFormatter::WIDTH_WIDE");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_open(locale, &status);
+    #endif
+
+    if (U_FAILURE(status)) {
+        intl_error_set(NULL, status, "Constructor failed", 0);
+        zend_throw_exception(IntlException_ce_ptr, "Constructor failed", 0);
+        RETURN_THROWS();
+    }
+}
+/* }}} */
+
+/* {{{ listformatter_format */
+PHP_METHOD(IntlListFormatter, format)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    zval *strings;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_ARRAY(strings)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (!LISTFORMATTER_OBJECT(obj)) {
+        intl_error_set(NULL, U_ILLEGAL_ARGUMENT_ERROR, "ListFormatter not properly constructed", 0);
+        RETURN_FALSE;
+    }
+
+    HashTable *ht = Z_ARRVAL_P(strings);
+    uint32_t count = zend_array_count(ht);
+    if (count == 0) {
+        RETURN_EMPTY_STRING();
+    }
+
+    const UChar **items = (const UChar **)safe_emalloc(count, sizeof(const UChar *), 0);
+    int32_t *itemLengths = (int32_t *)safe_emalloc(count, sizeof(int32_t), 0);
+    uint32_t i = 0;
+    zval *val;
+
+    ZEND_HASH_FOREACH_VAL(ht, val) {
+        zend_string *str_val;
+        
+        str_val = zval_get_string(val);
+        
+        // Convert PHP string to UTF-16
+        UChar *ustr = NULL;
+        int32_t ustr_len = 0;
+        UErrorCode status = U_ZERO_ERROR;
+        
+        intl_convert_utf8_to_utf16(&ustr, &ustr_len, ZSTR_VAL(str_val), ZSTR_LEN(str_val), &status);
+        zend_string_release(str_val);
+
+        if (U_FAILURE(status)) {
+            // We can't use goto cleanup because items and itemLengths are incompletely allocated
+            for (uint32_t j = 0; j < i; j++) {
+                efree((void *)items[j]);
+            }
+            efree(items);
+            efree(itemLengths);
+            intl_error_set(NULL, status, "Failed to convert string to UTF-16", 0);
+            RETURN_FALSE;
+        }
+        
+        items[i] = ustr;
+        itemLengths[i] = ustr_len;
+        i++;
+    } ZEND_HASH_FOREACH_END();
+
+    UErrorCode status = U_ZERO_ERROR;
+    int32_t resultLength;
+    UChar *result = NULL;
+
+    resultLength = ulistfmt_format(LISTFORMATTER_OBJECT(obj), items, itemLengths, count, NULL, 0, &status);
+
+    if (U_FAILURE(status) && status != U_BUFFER_OVERFLOW_ERROR) {
+        intl_error_set(NULL, status, "Failed to format list", 0);
+        RETVAL_FALSE;
+        goto cleanup;
+    }
+
+    // Allocate buffer and try again
+    status = U_ZERO_ERROR;
+    result = (UChar *)emalloc((resultLength + 1) * sizeof(UChar));
```

Comment:
Should be `safe_emalloc()`.

Reply:
quick glance, first and foremost maybe we need to check resultLength being < INT32_MAX (counting the +1 here) but I did not check ulistfmt_format implementation yet.

Reply:
Done :)